### ISSUE TITLE
utils: read_metadata can crash with a value error when numexpr is installed with pandas

### ISSUE THREAD

**Issue by huddlej** [nextstrain/augur#825]
**Current Behavior**  
[pandas recommends the installation of numexpr as a dependency](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html#recommended-dependencies) to speed up numerical operations for large datasets. When numexpr is installed, however, reading metadata with `augur.utils.read_metadata` can through a `ValueError` with the message of `unknown type object`. For example, the following code will produce a `ValueError`:

```python
from augur.utils import read_metadata
read_metadata('builds/h1n1pdm_cdc_egg_hi_2y/metadata.tsv')
```

The `ValueError` occurs when Augur attempts to [find duplicate records by strain name using the query method](https://github.com/nextstrain/augur/blob/73a07cfb39140a595d2a16ce1937a6bc54139db0/augur/util_support/metadata_file.py#L63). When Augur initially reads the metadata with pandas, [it parses the strain field as a dtype of "string"](https://github.com/nextstrain/augur/blob/73a07cfb39140a595d2a16ce1937a6bc54139db0/augur/util_support/metadata_file.py#L101). This dtype appears to be unsupported by numexpr. If we parse the strain field as an `object` type instead, the error does not occur. A more complete example follows:

```python
# Current logic. Fails with "unknown type object" error.
self.metadata["strain"].value_counts().reset_index().query("strain > 1")

# Same type of command with a different field that is parsed as an "object".
# Works as expected.
self.metadata["gender"].value_counts().reset_index().query("gender > 1400")

# Convert "strain" field to an "object" before applying current logic.
# Works as expected.
self.metadata["strain"].astype("object").value_counts().reset_index().query("strain > 1")
```

**Expected behavior**  
No error.

**How to reproduce**  

1. Install numexpr and augur: `python3 -m pip install nextstrain-augur numexpr`
2. Open a Python terminal
3. Load function: `from augur.utils import read_metadata`
4. Try to load metadata: `df = read_metadata("tests/functional/filter/metadata.tsv")`

**Possible solution**  
The error occurs in code where we try to identify duplicate records in a data frame using an older `read_metadata` function. The newer [`augur.io.read_metadata` function](https://github.com/nextstrain/augur/blob/73a07cfb39140a595d2a16ce1937a6bc54139db0/augur/io.py#L39) does not attempt to resolve duplicates and does not produce this error.

Rather than maintain an older function, we could address this error by replacing calls of the old function with the new function. [This is a list of where we use the old read_metadata function](https://github.com/nextstrain/augur/search?q=%22utils+import+read_metadata%22).

Alternately, we could maintain the older function and replace the current duplicate detection logic with more idiomatic logic for pandas using the `duplicated` method. For example, the following line should achieve the same result as the current logic to find duplicates:

```python
duplicates = self.metadata[self.metadata[self.key_type].duplicated()][self.key_type].values
```

This implementation avoids the problematic `query` method call that causes the numexpr error.

**Comment by victorlin** [comment]
> Rather than maintain an older function, we could address this error by replacing calls of the old function with the new function.

This is #813 😃

@huddlej have you started work on this issue yet? If not, I can work on it tomorrow.

**Comment by huddlej** [comment]
@victorlin I haven't started yet, so you're welcome to take this one. After running into this same error in the ncov workflow with some scripts there that rely on the old read_metadata, I think we should resolve this specific issue with the `duplicated()` approach described above and also separately resolve #813 by replacing references to the old read_metadata throughout Augur. We'll also need to update any external scripts in ncov and other workflows that call `augur.utils.read_metadata` directly. 😅

**Comment by TheBready** [comment]
Hey, I have the same issue with the latest conda version of augur. What is the conda ETA for the fix?

**Comment by huddlej** [comment]
Thanks for the +1 on this, @TheBready. We just released this in fix in 13.1.1 on PyPI and [the Conda package autobump just got merged](https://github.com/bioconda/bioconda-recipes/pull/32537). The updated Conda package should roll out soon.
